mesonbuild/build.py

```diff
@@ -1454,11 +1454,20 @@ def link_whole(self, target):
                     raise InvalidArguments(msg + ' This is not possible in a cross build.')
                 else:
                     mlog.warning(msg + ' This will fail in cross build.')
-            if isinstance(self, StaticLibrary):
-                # When we're a static library and we link_whole: to another static
-                # library, we need to add that target's objects to ourselves.
+            # When we're a static library and we link_whole: to another static
+            # library, we need to add that target's objects to ourselves.
+            # Otherwise add it to the link_whole_targets, but not both.
+            #
+            # We cannot use this with rust because of the way rust does crate
+            # search, or with fortran because of modules.
+            # TODO: it's like C++ with modules will not be able to suppor this
+            # optimization either
+            if (isinstance(self, StaticLibrary) and not
+                    ((self.uses_rust() or t.uses_rust()) or
+                     (self.uses_fortran() or t.uses_fortran()))):
```

Comment:
This is a bit involved and would be better if it was extracted to its own method with a descriptive name.